beater/waitintegration.go

```diff
@@ -0,0 +1,148 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.elastic.co/apm"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	"github.com/elastic/apm-server/elasticsearch"
+	"github.com/elastic/apm-server/kibana"
+	"github.com/elastic/go-elasticsearch/v7/esapi"
+)
+
+// waitForIntegration waits for the APM integration to be installed by querying Kibana,
+// or for the context to be cancelled.
+func waitForIntegration(
+	ctx context.Context,
+	kibanaClient kibana.Client,
+	esClient elasticsearch.Client,
+	interval time.Duration,
+	tracer *apm.Tracer,
+	logger *logp.Logger,
+) error {
+	logger.Info("waiting for integration package to be installed")
+	tx := tracer.StartTransaction("wait_for_integration", "init")
+	ctx = apm.ContextWithTransaction(ctx, tx)
+	var ticker *time.Ticker
+	for {
+		if ticker == nil {
+			ticker = time.NewTicker(interval)
+			defer ticker.Stop()
+		} else {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-ticker.C:
+			}
```

Comment:
I paused for a moment to think about why the code is written this way before realizing it's to send a request immediately, and then wait on subsequent loops. Could you write a short comment for Future Stuart:tm:?

Reply:
:+1: good call